[Operator] Add vdot

2niuhe opened the pull request:
### PR Category
 Operator

### Type of Change
New Feature

### Description
Implement vdot operator

- forward

### Issue

- Resolves #395 

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [ ] Change is fully covered by a UT.

### Performance

Test on T4, dtype complex64

```
vector-vdot-performance:
           size      Triton       Torch
0          16.0    0.032787    0.032345
1          32.0    0.107865    0.111628
2          64.0    0.241206    0.240000
3         128.0    0.475248    0.472906
4         256.0    0.950495    0.955224
5         512.0    1.910448    1.891626
6        1024.0    3.622642    3.728155
7        2048.0    7.078341    7.349282
8        4096.0   14.422535   14.355140
9        8192.0   25.869473   25.815127
10      16384.0   47.170825   45.850746
11      32768.0   76.800002   73.470855
12      65536.0  120.618400  114.707117
13     131072.0  173.375658  176.488333
14     262144.0  215.107227  218.453337
15     524288.0  252.872031  255.999991
16    1048576.0  315.076934  318.007282
17    2097152.0  360.500579  361.411758
18    4194304.0  385.269810  385.458646
19    8388608.0  401.240812  395.340943
20   16777216.0  410.307894  399.533637
21   33554432.0  414.000100  397.172244
22   67108864.0  416.713077  391.844561
23  134217728.0  417.201845  384.564819
```

2niuhe commented:
cc @StrongSpoon Need Review, Thanks

StrongSpoon commented on src/flag_gems/ops/vdot.py:
```diff
@@ -0,0 +1,206 @@
+from .. import runtime
+from ..utils import libentry
+
+tl_support_split = hasattr(tl, "split")
```
@Galaxy1458 could we add a check for these intrinsics supported only in triton v3?

StrongSpoon commented on src/flag_gems/ops/vdot.py:
```diff
@@ -0,0 +1,206 @@
+
+    base_offset = 2 * pid * BLOCK_SIZE + 2 * tl.arange(0, BLOCK_SIZE)
+
+    real_offset = base_offset[:, None] + tl.arange(0, 1)
```
no need to broadcast. I think 1d is okay.

2niuhe replied:
done

StrongSpoon commented on src/flag_gems/ops/vdot.py:
```diff
@@ -0,0 +1,206 @@
+    real_offset = base_offset[:, None] + tl.arange(0, 1)
+    imag_offset = real_offset + 1
+
+    mask = (base_offset < n_elements)[:, None]
```
ditto

2niuhe replied:
done

StrongSpoon reviewed (commented):
good job! I wonder if the performance data you show in this pr is latency or tflops.

StrongSpoon commented:
besides, I think that vdot_kernel_backwards_compatible also works on triton v3. have you compared the performance of two kernels?

2niuhe commented:
@StrongSpoon 

The test data above was obtained using `triton.testing.Benchmark`, and it refers to the throughput with the unit being GB/s.

--------

I compared the performance of the two kernels on the Nvidia T4, and found that the performance of the tl.split version is slightly faster.

However, I recently tested two kernels on other GPUs, such as the L4 GPU using Triton version 3.1, and observed that their performance is essentially the same, with no significant difference.

**Therefore, considering code simplicity, I have decided to remove the `tl.split` version.**

Tested on Nvidia L4 / Triton 3.1/ dtype=torch.complex64

- the split version kernel

```
vector-vdot-performance:
           size      Triton       Torch
0          16.0    0.046875    0.046875
1          32.0    0.093750    0.093750
2          64.0    0.187500    0.187500
3         128.0    0.428571    0.375000
4         256.0    0.750000    0.750000
5         512.0    1.333333    1.500000
6        1024.0    3.000000    2.666667
7        2048.0    5.333333    5.333333
8        4096.0    9.600000   10.666666
9        8192.0   19.200000   19.200000
10      16384.0   34.909091   34.909091
11      32768.0   63.999998   59.076924
12      65536.0  102.400003   96.000000
13     131072.0  153.600004  146.285715
14     262144.0  198.193551  192.000000
15     524288.0  231.849058  227.555555
16    1048576.0  236.307695  253.360834
17    2097152.0  252.061538  265.686491
18    4194304.0  267.130429  272.687926
19    8388608.0  301.546004  298.796351
20   16777216.0  341.629877  338.395869
21   33554432.0  366.634959  363.248045
22   67108864.0  380.010623  375.474815
23  134217728.0  387.619730  383.298167
```

- the common kernel
```
vector-vdot-performance:
           size      Triton       Torch
0          16.0    0.046875    0.041667
1          32.0    0.083333    0.083333
2          64.0    0.166667    0.166667
3         128.0    0.375000    0.333333
4         256.0    0.750000    0.666667
5         512.0    1.333333    1.333333
6        1024.0    2.666667    2.666667
7        2048.0    4.800000    5.333333
8        4096.0    9.600000    9.600000
9        8192.0   19.200000   19.200000
10      16384.0   34.909091   34.909091
11      32768.0   63.999998   59.076924
12      65536.0  102.400003   96.000000
13     131072.0  153.600004  146.285715
14     262144.0  198.193551  180.705879
15     524288.0  231.849058  227.555555
16    1048576.0  234.057145  246.994973
17    2097152.0  253.360834  265.686491
18    4194304.0  266.406514  272.310239
19    8388608.0  300.623865  298.796351
20   16777216.0  341.333321  338.395869
21   33554432.0  366.122895  363.080339
22   67108864.0  379.369039  375.206126
23  134217728.0  387.023602  383.181415
```

2niuhe commented on tests/test_blas_ops.py:
```diff
@@ -109,3 +115,32 @@ def test_accuracy_outer(M, N, dtype):
+
+
+@pytest.mark.vdot
+@pytest.mark.parametrize("M", [0, 1, 256, 2048, 65555])
```
done

2niuhe commented on tests/test_blas_ops.py:
```diff
@@ -109,3 +115,32 @@ def test_accuracy_outer(M, N, dtype):
+
+    with flag_gems.use_gems():
+        res_out = torch.vdot(inp1, inp2)
+    if dtype == torch.complex32:
```
done

2niuhe commented on src/flag_gems/ops/vdot.py:
```diff
@@ -0,0 +1,206 @@
+    ), f"Input tensors must have the same size. Got {input.size()} and {other.size()}."
+
+    inp = input.contiguous()
+    other = other.contiguous()
```
done

tongxin commented on src/flag_gems/ops/vdot.py:
```diff
@@ -0,0 +1,178 @@
+    pid = tl.program_id(0)
+    offset = pid * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+    mask = offset < n_elements
+
+    inp = tl.load(inp_ptr + inp_stride * offset, mask=mask).to(tl.float32)
+    other = tl.load(other_ptr + other_stride * offset, mask=mask).to(tl.float32)
+
+    out = tl.sum(inp * other)
+    tl.atomic_add(out_ptr, out)
```
If N is large, there will be severe contentions. It's more efficient to fix the number of CTAs and have each loop over the blocks in that case.

StrongSpoon replied:
agree

2niuhe replied:
I will make optimizations later.

2niuhe replied:
I attempted to fix the number of CTAs and have each loop over the blocks, as seen in [this commit](https://github.com/FlagOpen/FlagGems/commit/b7639d107e6fa328adae0c9d5208c75c801f2a62). After testing on the A100 with torch.cfloat, I found that there is a slight performance improvement when the number of loop iterations per block is well-matched with the number of blocks. However, it's challenging to determine a universal CTA count that works well across different GPUs. Additionally, when N is very large, it becomes difficult to optimize performance beyond the current implementation. 

Therefore, I believe that the contentions with atomic_add is not a significant issue in this context. If I come up with a better implementation in the future, I will submit a new PR to update it.


- Test1: Current Implementation, none fixed ctas

![image](https://github.com/user-attachments/assets/f7db70a4-be6b-400a-8e0f-79ca9f702df4)

- Test2: Max CTAs: 1 * SM number

![image](https://github.com/user-attachments/assets/09e8ecec-bf5f-418e-b995-b77fd8fd8a76)


- Test3: Max CTAs: 2 * SM number
![image](https://github.com/user-attachments/assets/ba79767f-7b7a-4ee6-9f7f-8b6b41c75148)


- Test4: Max CTAs: 4 * SM number

![image](https://github.com/user-attachments/assets/d1834593-4fbd-4a1a-9580-4a42193c7a55)

- Test5: Max CTAs: 6 * SM number

![image](https://github.com/user-attachments/assets/88144c8e-0882-4573-ad3a-5ad721f406de)

- Test6: Max CTAs: 8 * SM number

![image](https://github.com/user-attachments/assets/87cfae9b-8399-45dd-b9af-eec1ba816980)

-------

- BenchTest Code
 
```python
from flag_gems.ops import vdot
import torch
torch.set_default_device("cuda")
import triton

# benchtest


@triton.testing.perf_report(
    triton.testing.Benchmark(
        x_names=['size'],  # Argument names to use as an x-axis for the plot.
        x_vals=[2**i for i in range(4, 28, 1)],  # Different possible values for `x_name`.
        x_log=True,  # x axis is logarithmic.
        line_arg='provider',  # Argument name whose value corresponds to a different line in the plot.
        line_vals=['triton', 'torch'],  # Possible values for `line_arg`.
        line_names=['Triton', 'Torch'],  # Label name for the lines.
        styles=[('blue', '-'), ('green', '-')],  # Line styles.
        ylabel='GB/s',  # Label name for the y-axis.
        plot_name='vector-vdot-performance',  # Name for the plot. Used also as a file name for saving the plot.
        args={},  # Values for function arguments not in `x_names` and `y_name`.
    ))
def benchmark(size, provider):
    x = torch.randn(size, device='cuda', dtype=torch.cfloat)
    y = torch.randn(size, device='cuda', dtype=torch.cfloat)
    quantiles = [0.5, 0.2, 0.8]
    if provider == 'torch':
        ms, min_ms, max_ms = triton.testing.do_bench(lambda: torch.vdot(x, y), quantiles=quantiles)
    if provider == 'triton':
        ms, min_ms, max_ms = triton.testing.do_bench(lambda: vdot(x, y), quantiles=quantiles)
    gbps = lambda ms: 3 * x.numel() * x.element_size() * 1e-9 / (ms * 1e-3)
    return gbps(ms), gbps(max_ms), gbps(min_ms)

benchmark.run(print_data=True, show_plots=True)
```

StrongSpoon reviewed (approved):
lgtm
